pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,349 @@
+}
+
+func FromSecret(secret *corev1.Secret) (*LEUser, error) {
+	block, _ := pem.Decode(secret.Data["privateKey"])
+	x509Encoded := block.Bytes
+	privateKey, err := x509.ParseECPrivateKey(x509Encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	var reg registration.Resource
+	if err := json.Unmarshal(secret.Data["registration"], &reg); err != nil {
+		return nil, err
+	}
+
+	return &LEUser{
+		Name:         string(secret.Data["name"]),
+		Email:        string(secret.Data["email"]),
+		Registration: &reg,
+		Key:          privateKey,
+		URL:          string(secret.Data["url"]),
+	}, nil
+}
+
+func (u *LEUser) GetEmail() string {
+	return u.Email
+}
+func (u LEUser) GetRegistration() *registration.Resource {
+	return u.Registration
+}
+func (u *LEUser) GetPrivateKey() crypto.PrivateKey {
+	return u.Key
+}
+
+func (u *LEUser) Register() error {
+	if u.Name == "" || u.Email == "" {
+		return fmt.Errorf("not registering LE User: missing name or email")
+	}
+
+	if u.URL == "" {
+		return fmt.Errorf("not registering LE User: missing URL")
+	}
+
+	if u.Key == nil {
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return err
+		}
+		u.Key = key
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return err
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+	if err != nil {
+		return err
+	}
+	u.Registration = reg
+
+	logrus.Infof("registered LE User: %s", u.Name)
+
+	return nil
+
+}
+
+func (u *LEUser) ToSecret() (*corev1.Secret, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not saving LE User: missing registration")
+	}
+	x509Encoded, _ := x509.MarshalECPrivateKey(u.Key.(*ecdsa.PrivateKey))
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509Encoded})
+
+	reg, err := json.Marshal(u.Registration)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.LESecretName,
+			Namespace: system.Namespace,
+			Labels: map[string]string{
+				labels.AcornManaged: "true",
+			},
+		},
+		Data: map[string][]byte{
+			"name":         []byte(u.Name),
+			"email":        []byte(u.Email),
+			"privateKey":   pemEncoded,
+			"registration": reg,
+			"url":          []byte(u.URL),
+		},
+	}, nil
+}
+
+func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certificate.Resource, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not generating LE cert: missing registration")
+	}
+
+	if domain == "" || token == "" {
+		return nil, fmt.Errorf("not generating LE cert: missing domain or token")
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)
+
+	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
```

Comment:
As mentioned in slack, I ran into a lot of trouble with my local DNS resolver trying to do the "pre check" for DNS. Not sure if you caught that bit of the convo, but this client attempts to ensure that the TXT record is there locally before telling let's encrypt to do it. The problem is that local dns resolvers can act really funky and for me, they never succeed. We agreed that if we just wait 60 seconds (the documented time for dns to proprogate throughout route53), we can just skip the local check entirely.

Here is the diff that allowed me to achieve this locally:
```
diff --git a/pkg/controller/dns/letsencrypt.go b/pkg/controller/dns/letsencrypt.go
index c5af25f..a17b690 100644
--- a/pkg/controller/dns/letsencrypt.go
+++ b/pkg/controller/dns/letsencrypt.go
@@ -21,6 +21,7 @@ import (
 	"github.com/acorn-io/baaah/pkg/router"
 	"github.com/go-acme/lego/v4/certcrypto"
 	"github.com/go-acme/lego/v4/certificate"
+	"github.com/go-acme/lego/v4/challenge/dns01"
 	"github.com/go-acme/lego/v4/lego"
 	"github.com/go-acme/lego/v4/registration"
 	"github.com/sirupsen/logrus"
@@ -143,6 +144,10 @@ func (u *LEUser) ToSecret() (*corev1.Secret, error) {
 	}, nil
 }

+func noOpCheck(_, _, _ string, _ dns01.PreCheckFunc) (bool, error) {
+	return true, nil
+}
+
 func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certificate.Resource, error) {
 	if u.Registration == nil {
 		return nil, fmt.Errorf("not generating LE cert: missing registration")
@@ -163,7 +168,7 @@ func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certi

 	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)

-	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
+	if err := client.Challenge.SetDNS01Provider(dnsProvider, dns01.WrapPreCheck(noOpCheck)); err != nil {
 		return nil, err
 	}

diff --git a/pkg/controller/dns/provider.go b/pkg/controller/dns/provider.go
index 93c1b20..612167d 100644
--- a/pkg/controller/dns/provider.go
+++ b/pkg/controller/dns/provider.go
@@ -40,7 +40,7 @@ func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
 }

 func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
-	return 30 * time.Second, 5 * time.Second
+	return 3 * time.Minute, 1 * time.Minut
 }

```